### PR TITLE
Fix weight calculator

### LINKED ISSUE #4695: Checkout broken if weight calculator used in a payment or shipping method AND a product in the cart has unit_value ZERO
## Description
If a shipping or payment method uses the weight calculator and a variant being checked out has variant unit different from weight (items for example) and has the unit value set to zero by the user. The checkout button wont work and will not show an error to the user.

## Expected Behavior
Checkout works and the product is included in the order.
The shipping method cost is calculated in some default way: what does zero unit_value mean? "zero items", "zero kilos", "zero liters" is a bit non-sense. The cost should probably default to zero in these cases.

## Actual Behaviour
Final checkout button does nothing. No error is shown.

## Steps to Reproduce
Setup a shipping method with a weight based calculator.
Create a product with variant_unit items and unit_value 0.
Add to OC
Checkout the product.
Bang.

## Workaround
Change the shipping method to something else.
Change the product unit value to something else.

## Severity
This was S2 until we discovered what it was. It's now an S3 I think. I got two cases in Katuma where users are setting unit value to zero. And weight based calculators are very common. This will happen again for sure.

## Your Environment
Katuma

* Version used:
* Browser name and version: firefox latest
* Operating System and version (desktop or mobile): osx catalina

## Possible Fix
Bugsnag - https://app.bugsnag.com/katuma/katuma/errors/5de6470bd3c5fe001a1ae122?filters[event.since][0]=30d&filters[error.status][0]=open

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Closes #4695

Fixes issue where weight calculator was causing fatal errors by returning `BigDecimal::NaN` in some cases.

In the case where `variant.product.unit_type` was `items`, and `variant.unit_value` was set to `0`, and `variant.weight` was not set; we were **dividing by zero** in one of the calculations.

#### What should we test?

Checkout should work when: 
- one of the products in the order uses "items" as the type and has unit_value "0".
- the shipping method uses the weight calculator

#### Release notes

Fixed problematic edge case for shipping methods using weight calculator

Changelog Category: Fixed

### CODE REVIEW

File: app/models/calculator/weight.rb [hunk 1]
```diff
@@ -54,6 +54,8 @@ def weight_per_final_weight_volume(line_item)
     #    Customer ends up getting 350mL (line_item.final_weight_volume) of wine
     #      that represent 2.8 (quantity_implied_in_final_weight_volume) glasses of wine
     def quantity_implied_in_final_weight_volume(line_item)
+      return 0 if line_item.variant.unit_value.zero? # Don't divide by zero ;)
```
Comment (luisramos0) [inline app/models/calculator/weight.rb]: you could default to 1 here, that would mean the variant.weight field would be used. this way the weight will be zero even if the weight field on the variant is defined.

Comment (luisramos0) [reply app/models/calculator/weight.rb]: no, better, I think we should default to line_item.quantity.
it would read "if unit_value is zero, the quantity_implied_in_final_weight_volume cannot be calculated so we use the line_item.quantity

Comment (Matt-Yorkley) [reply app/models/calculator/weight.rb]: We're checking for the presence of `final_weight_volume`, but it always appears to be present and defaults to zero. I don't think we should use the volume calculations at all when `variant_unit` is "items"...?

Comment (Matt-Yorkley) [reply app/models/calculator/weight.rb]: It still returns zero anyway, because the calculator sets `:preferred_per_kg` to `0.0`, and we multiply the result by that... :see_no_evil:

File: app/models/calculator/weight.rb [hunk 2]
```diff
@@ -54,11 +54,14 @@ def weight_per_final_weight_volume(line_item)
     #    Customer ends up getting 350mL (line_item.final_weight_volume) of wine
     #      that represent 2.8 (quantity_implied_in_final_weight_volume) glasses of wine
     def quantity_implied_in_final_weight_volume(line_item)
+      return weight_per_variant(line_item) if line_item.variant.unit_value.zero?
+
       (1.0 * line_item.final_weight_volume / line_item.variant.unit_value).round(3)
     end
 
     def final_weight_volume_present?(line_item)
```
Comment (Matt-Yorkley) [inline app/models/calculator/weight.rb]: This check was always returning `true`; the default value is a `BigDecimal` object with value `0.0`...

**Comment by Matt-Yorkley** [comment]
@luisramos0 I added an extra test, changed some key bits of the logic and extracted the "never divide by zero" bit into a separate commit. Needs a careful re-review...

File: app/models/calculator/weight.rb (hunk 1 above)
Comment (luisramos0) [reply app/models/calculator/weight.rb]: final_weight_volume is not always present, for example for subscriptions. the check fixed a bug if you look into the commit history.

it's a good question: I think you can change the final weight volume of line items even if the variant_unit is items... in these cases final_weight_volume  becomes variant.unit_value time line_item quantity.
So, 
product.variant_unit items
variant.display_name "5 roses"
variant.unit_value 5
variant.weight 20g
line_item.quantity 2
line_item.final_weight_volume 10 (variant.unit_value * line_item.quantity)

for this line item weight calculator should calculate 40grams

but if user changes line_item.final_weight_volume from 10 to 8 (8 roses were actually delivered) the weight calculator will compute correctly 32grams :+1:  20 * (8/5)

So, I think it's good as is.

Comment (luisramos0) [reply app/models/calculator/weight.rb]: the field name should not be final_weight_volume
as I demonstrate above it's actually final_weight_volume_units, in this case final_units.

File: app/models/calculator/weight.rb (hunk 2 above)
Comment (luisramos0) [reply app/models/calculator/weight.rb]: I added it to fix a bug
e9e6aa77d88ebc6d93277d85bbc2e4eca1f558d2

File: app/models/calculator/weight.rb [hunk 3]
```diff
@@ -57,8 +57,13 @@ def quantity_implied_in_final_weight_volume(line_item)
...
     end
 
+    def use_final_weight_volume?(line_item)
+      final_weight_volume_present?(line_item) && !line_item.variant.unit_value.zero?
+    end
+
     def final_weight_volume_present?(line_item)
-      line_item.respond_to?(:final_weight_volume) && line_item.final_weight_volume.present?
+      line_item.respond_to?(:final_weight_volume) && line_item.final_weight_volume.present? &&
+        !line_item.final_weight_volume.zero?
```
Comment (luisramos0) [inline app/models/calculator/weight.rb]: looking at the DB I suspect a zero final_weight_volume is actually meaningful (set by user) and we should use it... I'd not change that behavior.

File: app/models/calculator/weight.rb [hunk 4]
```diff
@@ -57,8 +57,13 @@ def quantity_implied_in_final_weight_volume(line_item)
       (1.0 * line_item.final_weight_volume / line_item.variant.unit_value).round(3)
     end
 
+    def use_final_weight_volume?(line_item)
+      final_weight_volume_present?(line_item) && !line_item.variant.unit_value.zero?
```
Comment (luisramos0) [inline app/models/calculator/weight.rb]: I dont think this is correct. If variant_unit is weight, you dont need unit_value to use final_weight_volume.
I think the previous version of the code is better where we just change quantity_implied_in_final_weight_volume

**Comment by Matt-Yorkley** [comment]
Version 3 is ready for review. I played a lot with additional test cases and it was a minefield of `NaN` errors...

**Comment by Matt-Yorkley** [comment]
Also I noticed the other tests in this spec file all use `:build` in the line item factories instead of `:create`, which skips the callback that calculates and saves the `final_weight_volume`...

**Comment by Matt-Yorkley** [comment]
Oops... there was an extra `end` in one of your commit lines that broke the build. It's fixed now :+1:

**Comment by luisramos0** [comment]
oh, I am sorry! I must have messed it up while splitting the code in two commits...

File: app/models/calculator/weight.rb [hunk 5]
```diff
@@ -54,6 +54,8 @@ def weight_per_final_weight_volume(line_item)
     #    Customer ends up getting 350mL (line_item.final_weight_volume) of wine
     #      that represent 2.8 (quantity_implied_in_final_weight_volume) glasses of wine
     def quantity_implied_in_final_weight_volume(line_item)
+      return line_item.quantity if line_item.variant.unit_value.zero?
```
Comment (mkllnk) [inline app/models/calculator/weight.rb]: I looked at au-prod:

```rb
Spree::Variant.where(unit_value: 0).count
# => 38

Spree::Variant.where(unit_value: nil).count
# => 630
```

I think we can catch these cases as well:

```suggestion
      return line_item.quantity if line_item.variant.unit_value.to_i.zero?
```

What do you think? Is that case covered somewhere else?

Comment (mkllnk) [reply app/models/calculator/weight.rb]: I'm not sure if this is a valid case but we do have data like this:
```rb
item = Spree::LineItem.where("final_weight_volume > 0").joins(:variant).where("spree_variants.unit_value is null").last
Calculator::Weight.new.compute(item)
# TypeError: nil can't be coerced into BigDecimal
```

**Comment by filipefurtad0** [comment]
Hi @Matt-Yorkley,

I have been trying to reproduce the bug - but it does not seem to be silent anymore. So, now one gets a warning "The checkout failed. Please let us know so that we can process your order." 

In any case, after following the test cases - all seems to work fine. Checkout is now possible, and the delivery costs are set by default to zero, on the delivery "items" as units. Also, I checked that everything works normally, and that the calculator works normally for weight items.

Good to go.

**Comment by Matt-Yorkley** [comment]
Thanks @filipefurtad0! Awesome testing :muscle:
